### ISSUE TITLE
Generation of dependency files in external library does depend on content of file itself

### ISSUE THREAD

**Issue by christianlupus** [pgf-tikz/pgf#866]
**Brief outline of the bug**

The externalization library produces wrong dependency files in mode `list and make`.

I am not relying on the update checking mechanism using md5 as it involves a compile run of the main file to update the md5 sums. Normally, I keep it active, but while developing on the images it is much more convenient for me if I can simply call `make images` and all required images are rebuilt without the whole project. This is especially true for beamer presentations with overlays where not only one image but a series might be built.

**Minimal working example (MWE)**

The MWE consists of three files in total as I need to focus on the dependency file generation by the tikz external library.

```latex
% mwe.tex
\documentclass{article}
\usepackage{tikz}
\usetikzlibrary{external}
\tikzexternalize[mode=list and make,up to date check=simple]

\newcommand{\importtikz}[1]{
	\typeout{Externalizing #1}
	\tikzpicturedependsonfile{#1} % <--- Occurrence (1) of tikzpicturedependsonfile
	\input{#1}
}


\begin{document}
\importtikz{imgb.tikz}
\end{document}
```

The idea here is to just import a tikz external file into the document by a self-written macro. This will do other things as well in the real world but here this is sufficient.

```latex
% imgb.tikz

% Occurrences of tikzpicturedependsonfile (2)
% \tikzpicturedependsonfile{imga.tikz}
% \tikzpicturedependsonfile{imgb.tikz}
\begin{tikzpicture}
% Occurances of tikzpicturedependsonfile (3)
% \tikzpicturedependsonfile{imga.tikz}
% \tikzpicturedependsonfile{imgb.tikz}
\draw (0,0) circle (2cm);
\end{tikzpicture}
```

Just some arbitrary image code. Please note that the dependency declarations are commented out for now.

Further, I need a file `imga.tikz`. You can simply copy the code over from `imgb.tikz`.

**Now do the following steps**

1. Compile with `pdflatex` and `make` as is
1. Uncomment any occurrence in (2) and recompile
1. Recomment the occurrences in (2) and uncomment the occurrence in (3) for `imgb` and recompile
1. Uncomment the occurrence in (3) for `imga` and recompile
1. Recomment all occurrences in (3) and recompile

**What I expect**

Each time the `mwe.*` files are updated and the `make` command will build a `mwe-figure0.*` bundle. I am focusing now on the `mwe-figure0.dep` file.

1. After the first run it should contain a line rendering `mwe-figure0.pdf` dependent on `imgb.tikz`.
1. The same after the second (maybe doubled but ok)
1. Also the same
1. An additional dependency on `imga.tikz`
1. The **exact** same result as after step 1

**What I get**

1. After the first step the file `mwe-figure0.dep` is empty
1. No change in the file content, it stays empty
1. Now a dependency is added as expected
1. The additional dependency is added
1. The dependencies keep their state: It is still dependent on both files

You can test any combination in (3), compile it, and comment all in (3): You will always get the same dependencies as if the file was not touched at all. I cannot tell, if this is really the case or not.

**Additional comments**

The occurrences in (1) and (2) are somehow redundant. I just wanted to avoid that it needs to be exactly before the `tikzpicture` itself.

In fact, there are 2 sub concerns that I assume to be related (thus one bug report):
1. The dependency is only taken into account if defined inside the corresponding `tikzpicture`. According to [the manual](https://github.com/pgf-tikz/pgf/blob/master/doc/generic/pgf/text-en/pgfmanual-en-library-external.tex#L440-L444) it should also work for the next picture.
1. If no dependencies are explicitly specified, the `dep` file is not altered at all.

**Comment by hmenke** [comment]
I'm pretty sure that the generation of the dependency files *is* deterministic.  It's just wrong.

**Comment by christianlupus** [comment]
Yeah, I should have written *appears non-deterministic to the user without further research*. I was first a bit stunned as I assumed the file content would only depend on the TeX file but not the state of the `dep` file itself.

**Comment by hmenke** [comment]
It also doesn't “appear non-deterministic” to the user.  Non-determinism has an entirely different meaning.

**Comment by christian-wolf-eks** [comment]
I checked this issue again (partially as I forgot about the issue in the meantime). In fact, i think with the current installation (I have pgf 3.1.10 installed from MikTex) the problem is a ISO layer 8 problem 😳.

In fact, I found that the location of the macro `\tikzpicturedependsonfile` makes a difference on where the dependency is applied:

When used **inside the tikzpicture**, the corresponding `.dep` file is used to communicate the dependencies to `make`. Thus, when building a single (externalized) image, it creates a set of files (`.pdf`, `.dep`, `.dpth`, `.log`, ...). This works as expected.

When using **outside the actual picture**, the information is not stored in the `.dep` file. Instead, the corresponding line is printed into the `.makefile` of the main compiler run. I did oversee this as the makefile was rather lengthy and I did not check every single line of every generated file.

This does in fact make sense as then the responsibilities are clear: the `.makefile` is generated by the main latex run and the `.dep` files are generated by the individual runs. This way, the files are not mutually overwritten.

As this issue seems no problem o be, I'd close this issue here.
